client/client.go

```diff
@@ -22,146 +23,153 @@ const (
-	// Base URL of the service (https://keys.sylabs.io is used if not supplied).
-	BaseURL string
-	// Auth token to include in the Authorization header of each request (if supplied).
-	AuthToken string
-	// User agent to include in each request (if supplied).
-	UserAgent string
-	// HTTPClient to use to make HTTP requests (if supplied).
-	HTTPClient *http.Client
-}
+// errUnsupportedProtocolScheme is returned when an unsupported protocol scheme is encountered.
+var errUnsupportedProtocolScheme = errors.New("unsupported protocol scheme")
 
-// DefaultConfig is a configuration that uses default values.
-var DefaultConfig = &Config{}
+// ErrTLSRequired is returned when an auth token is supplied with a non-TLS BaseURL.
+var ErrTLSRequired = errors.New("TLS required when auth token provided")
 
-// PageDetails includes pagination details.
-type PageDetails struct {
-	// Maximum number of results per page (server may ignore or return fewer).
-	Size int
-	// Token for next page (advanced with each request, empty for last page).
-	Token string
-}
-
-// Client describes the client details.
-type Client struct {
-	// Base URL of the service.
-	BaseURL *url.URL
-	// Auth token to include in the Authorization header of each request (if supplied).
-	AuthToken string
-	// User agent to include in each request (if supplied).
-	UserAgent string
-	// HTTPClient to use to make HTTP requests.
-	HTTPClient *http.Client
-}
+// normalizeURL normalizes rawURL, translating HKP/HKPS schemes to HTTP/HTTPS respectively, and
+// ensures the path component is terminated with a separator.
+func normalizeURL(rawURL string) (*url.URL, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, err
+	}
 
-// normalizeURL normalizes the scheme of the supplied URL. If an unsupported scheme is provided, an
-// error is returned.
-func normalizeURL(u *url.URL) (*url.URL, error) {
 	switch u.Scheme {
 	case schemeHTTP, schemeHTTPS:
-		return u, nil
+		break
 	case schemeHKP:
 		// The HKP scheme is HTTP and implies port 11371.
-		newURL := *u
-		newURL.Scheme = schemeHTTP
+		u.Scheme = schemeHTTP
 		if u.Port() == "" {
-			newURL.Host = net.JoinHostPort(u.Hostname(), "11371")
+			u.Host = net.JoinHostPort(u.Hostname(), "11371")
 		}
-		return &newURL, nil
 	case schemeHKPS:
 		// The HKPS scheme is HTTPS and implies port 443.
-		newURL := *u
-		newURL.Scheme = schemeHTTPS
-		return &newURL, nil
+		u.Scheme = schemeHTTPS
 	default:
-		return nil, fmt.Errorf("unsupported protocol scheme %q", u.Scheme)
+		return nil, fmt.Errorf("%w %s", errUnsupportedProtocolScheme, u.Scheme)
+	}
+
+	// Ensure path is terminated with a separator, to prevent url.ResolveReference from stripping
+	// the final path component of BaseURL when constructing request URL from a relative path.
+	if !strings.HasSuffix(u.Path, "/") {
+		u.Path += "/"
 	}
+
+	return u, nil
 }
 
-const defaultBaseURL = "https://keys.sylabs.io/"
+// Client describes the client details.
+type Client struct {
```

Comment:
@tri-adam Would it be possible to also pass a callback function in order to grab the body content response or copy the content to retrieve if after the request ? In case of success it could be helpful to display content to give some hint. As for error by example in `pkg/sypgp/sypgp.go` there is an hardcoded hint to guide users about authentication, as this is something specific to an endpoint the keyserver could provide this hint instead via the JSON error response with a hint field maybe ?

Reply:
Hi @cclerget, apologies for the slow response!

I'm making the assumption you're referring solely to [(*Client).PKSLookup()](https://github.com/tri-adam/scs-key-client/blob/aecd579e103f9fcdf36dbe1b95394cfd6f44b3a4/client/pks.go#L86-L90), as that's the relevant API for the specific part of the code you reference in the `sypgp` package.

I'm not sure there is much use to have an alternative way to get a copy of the body when the request was _successful_, since it is already returned in its entirety in this case.

When the request is _unsuccessful_, it will return a [(*github.com/sylabs/scs-key-client/client).HTTPError](https://github.com/tri-adam/scs-key-client/blob/aecd579e103f9fcdf36dbe1b95394cfd6f44b3a4/client/error.go#L16-L20), which a caller could then use `errors.As()` and [(*HTTPError).Code()](https://github.com/tri-adam/scs-key-client/blob/aecd579e103f9fcdf36dbe1b95394cfd6f44b3a4/client/error.go#L22-L23) to obtain the HTTP status code. Since the HKP draft specification really doesn't define error/hint formats, this is likely a good level for Singularity to do its logic at in terms of whether authentication is necessary.

The use case you describe about providing hints was actually the inspiration for the [github.com/sylabs/json-resp](https://pkg.go.dev/github.com/sylabs/json-resp) package. Since all Sylabs HKP servers use that format, this module will also return a wrapped [jsonresp.Error](https://pkg.go.dev/github.com/sylabs/json-resp#Error) when the response contains that format, making it possible to return more detailed help to the user via the `Message` field.

But, back to your question... if the ask is to handle some other type of error format, I would suggest that exposing the response body via a `Body()` method on the [(*github.com/sylabs/scs-key-client/client).HTTPError](https://github.com/tri-adam/scs-key-client/blob/aecd579e103f9fcdf36dbe1b95394cfd6f44b3a4/client/error.go#L16-L20) would be the idiomatic way to do so. The caller could then implement any sort of custom logic they choose to mine any clues from the content of the response.

Let me know what you think, and thanks for taking the time to provide feedback!

Reply:
Hi @tri-adam , I'm referring to `PKSAdd` as there is key server like Hagrid https://keys.openpgp.org/about/api by example which are providing some hint as a result of a push because it does additional verification regarding the key being uploaded and it would be nice to return the response message on CLI then a user could proceed with verification by following instructions returned by the server without being misguided by a message like : your key has been successfully pushed. So just returning nil from `PKSAdd` as a success can't cover those cases.
That's not really related to error as this is pretty much covered and can be done from singularity at this level

Reply:
Ah, that makes sense, thanks for the clarification. I've opened #47 to track that. I think it's best kept separate from this refactoring, given the scope of this is already quite large.